example/src/main/java/org/mozilla/fxa_data/example/KotlinExampleActivity.kt

```diff
@@ -0,0 +1,210 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fxa_data.example
+
+import android.content.Intent
+import android.os.Bundle
+import android.support.annotation.StringRes
+import android.support.v7.app.AppCompatActivity
+import android.support.v7.widget.DividerItemDecoration
+import android.support.v7.widget.LinearLayoutManager
+import android.support.v7.widget.RecyclerView
+import android.view.View
+import kotlinx.android.synthetic.main.container_explanation.*
+import kotlinx.android.synthetic.main.container_loading.*
+import kotlinx.android.synthetic.main.container_show_fx_data.*
+import org.mozilla.fxa_data.FirefoxData
+import org.mozilla.fxa_data.FirefoxDataException
+import org.mozilla.fxa_data.login.FirefoxDataLoginManager
+import org.mozilla.fxa_data.download.FirefoxDataClient
+import org.mozilla.fxa_data.download.HistoryRecord
+import org.mozilla.fxa_data.download.FirefoxDataResult
+import java.lang.Exception
+
+import java.lang.ref.WeakReference
+
+/**
+ * An example Activity that which displays a user's history in a [RecyclerView] after they log in.
+ * If they're already logged in, the data will just be shown. Once the data is shown, the user can click
+ * "Sign out" to sign out.
+
+ * This Activity will switch between four states:
+ * - Prompt for sign in
+ * - Loading
+ * - Error logging in or fetching Firefox data
+ * - Display user's history.
+
+ * This is a fairly realistic use case. For a simpler example, see [SimpleExampleActivity].
+ */
+public class KotlinExampleActivity : AppCompatActivity() {
+
+    private enum class UIState {
+        SIGN_IN_PROMPT, LOADING, ERROR, SHOW_DATA
+    }
+
+    private lateinit var loginManager: FirefoxDataLoginManager
+    private lateinit var historyAdapter: FirefoxHistoryAdapter
+
+    private var isWaitingForCallback = false
+
+    private val containerViews by lazy {
+        listOf(container_explanation,
+                container_loading,
+                container_fx_data)
+    }
+
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_firefox_data_in_recycler_view_example)
+
+        supportActionBar?.title = KotlinExampleActivity::class.simpleName
+
+        initFirefoxDataView()
+
+        // Keep a reference to the login manager - it's the entry point for Firefox data.
+        loginManager = FirefoxData.getLoginManager(this)
+    }
+
+    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
+        super.onActivityResult(requestCode, resultCode, data)
+
+        if (data == null) {
```

Comment:
I don't think we need this clause – the loginManager gracefully handles `data == null`. Additionally, we don't know for sure that this intent is even for the login manager and the intent could be sent for elsewhere with null data.

Reply:
Oh, I see the commit title now. I think if we want to add features, we should do it to both example activities.

Also, I think the proper way to do this in this library would be to restore the prompt in the onError callback since you don't know for sure who is going to be calling into `onActivityResult`.

Reply:
The problem I want to solve here is when the user press back button at the login page. The intent data will be null, and loginManager won't handle [the situation when intent is null](https://github.com/mozilla-mobile/FirefoxData-android/blob/master/download/src/main/java/org/mozilla/fxa_data/login/FirefoxDataWebViewLoginManager.java#L70). Thus the UI state won't get back to SIGN_IN_PROMPT. 

Today when I test the app, I had ["Account failed to advance to Married state for unknown reason" ](https://github.com/mozilla-mobile/FirefoxData-android/blob/master/download/src/main/java/org/mozilla/fxa_data/login/FirefoxDataWebViewLoginManager.java#L122) Would you please advise? Thanks!

Reply:
Hey @cnevinc – I'm on PTO tomorrow but I'll look at this on Monday – thanks!

fwiw, it's entirely possible I didn't think of this case for my library and it's actually a flaw in its design. :( But I'll see soon.

Reply:
Hey @cnevinc – this is a bug in the library: I filed #31 and put it into the v0.1 milestone (i.e. first real release). Roughly, it's that we should be returning an intent with `RESULT_CANCELED` but instead null is being returned.

I think if you remove this clause and we fix the bug, it'll be fine. Also, if you can fix the naming nit.

Sorry again for the delay – this project isn't prioritized so I'm struggling to put it into my  workload from day-to-day. For what it's worth, if you want to do more Kotlin coding, we're starting to use it in Focus so we can probably find you some issues there that we'll be happy to review quickly. :) Let me know!